thetis/interpolation.py

```diff
@@ -0,0 +1,634 @@
+        fsx = Function(self.function_space).interpolate(xy[0]).dat.data_with_halos
+        fsy = Function(self.function_space).interpolate(xy[1]).dat.data_with_halos
+
+        mesh_lonlat = []
+        for node in range(len(fsx)):
+            lat, lon = to_latlon(fsx[node], fsy[node])
+            mesh_lonlat.append((lon, lat))
+        self.mesh_lonlat = np.array(mesh_lonlat)
+
+        self._initialized = False
+
+    def _get_subset_nodes(self, grid_x, grid_y, target_x, target_y):
+        """
+        Retuns grid nodes that are necessary for intepolating onto target_x,y
+        """
+        orig_shape = grid_x.shape
+        grid_xy = np.array((grid_x.ravel(), grid_y.ravel())).T
+        target_xy = np.array((target_x.ravel(), target_y.ravel())).T
+        tri = qhull.Delaunay(grid_xy)
+        simplex = tri.find_simplex(target_xy)
+        vertices = np.take(tri.simplices, simplex, axis=0)
+        nodes = np.unique(vertices.ravel())
+        nodes_x, nodes_y = np.unravel_index(nodes, orig_shape)
+
+        # x and y bounds for reading a subset of the netcdf data
+        ind_x = slice(nodes_x.min(), nodes_x.max() + 1)
+        ind_y = slice(nodes_y.min(), nodes_y.max() + 1)
+
+        return nodes, ind_x, ind_y
+
+    def _create_interpolator(self, lat_array, lon_array):
+        """
+        Create compact interpolator by finding the minimal necessary support
+        """
+        self.nodes, self.ind_lon, self.ind_lat = self._get_subset_nodes(
+            lon_array,
+            lat_array,
+            self.mesh_lonlat[:, 0],
+            self.mesh_lonlat[:, 1]
+        )
+
+        subset_lat = lat_array[self.ind_lon, self.ind_lat].ravel()
+        subset_lon = lon_array[self.ind_lon, self.ind_lat].ravel()
+        subset_lonlat = np.array((subset_lon, subset_lat)).T
+        self.interpolator = GridInterpolator(subset_lonlat, self.mesh_lonlat)
+        self._initialized = True
+
+        # debug: plot subsets
+        # import matplotlib.pyplot as plt
+        # plt.plot(grid_lon_full, grid_lat_full, 'k.')
+        # plt.plot(grid_lonlat[:, 0], grid_lonlat[:, 1], 'b.')
+        # plt.plot(self.mesh_lonlat[:, 0], self.mesh_lonlat[:, 1], 'r.')
+        # plt.show()
+
+    @abstractmethod
+    def interpolate(self, data):
+        """
+        Calls the interpolator object
+        """
+        return self.interpolator(data)
+
+
+class NetCDFLatLonInterpolator2d(SpatialInterpolator2d):
+    """
+    Interpolates netCDF data on local 2D unstructured mesh
+
+    This routine returns the data in numpy arrays.
+
+    Usage:
+
+    .. code-block:: python
+        fs = FunctionSpace(...)
+        myfunc = Function(fs, ...)
+        ncinterp2d = NetCDFLatLonInterpolator2d(fs, to_latlon)
+        val1, val2 = ncinterp2d.interpolate(nc_filename, ['var1', 'var2'], 10)
+        myfunc.dat.data_with_halos[:] = val1 + val2
+
+    """
+    def interpolate(self, nc_filename, variable_list, itime):
```

Comment:
This kind of feels wrong in that the method signature doesn't match the abstract class.  Would it not make more sense to  push the setup this function does into the constructor?  After all, this method signature suggests that you can use one `NetCDFLatLonInterpolator2d` to interpolate into some specified function space from multiple different netcdf files.  But that isn't the case, because the interpolator object is made from the first netcdf file that is passed in.

Reply:
I have changed the abstract classes now such that `interpolate` method always takes a filename.

I'd like to avoid defining the filename in the constructor because it's generally not known at that time. If the data set is divided into multiple files, the time search objects will seek for the correct file. Then, it's then easy to pass the filename and time index to the spatial interpolator who will initialize itself if needed. The assumption, of course, is all the files have the same spatial grid.

Reply:
OK, that makes sense.